methods_enumerable.rb

```diff
@@ -0,0 +1,159 @@
+module Enumerable
+  def my_each
+    for item in self
+      yield item
+    end
+  end
+
+  def my_each_with_index
+    i = 0
+    for item in self
+      yield(item, i)
+      i += 1
+    end
+  end
+
+  def my_select
+      res = []
+      my_each { |v| res.push(v) if yield(v) }
+      return res
+  end
+
+
+  def my_all?(pattern = nil)
+    if block_given?
+      my_each { |v| return false unless yield(v) }
+    elsif !pattern.nil?
+      if pattern.is_a?(Regexp)
+        my_each { |v| return false unless pattern =~ v.to_s }
+      else
+        my_each { |v| return false unless v.is_a?(pattern) }
+      end
+    else
+      my_each { |v| return false unless v }
+    end
+    true
+  end
+
+  def my_any?(pattern = nil)
+    if block_given?
+      my_each { |v| return true if yield(v) }
+    elsif !pattern.nil?
+      if pattern.is_a?(Regexp)
+        my_each { |v| return true if pattern =~ v.to_s }
+      else
+        my_each { |v| return true if v.is_a?(pattern) }
+      end
+    else
+      my_each { |v|  v != nil ?  true : false }
+    end
+    false
+  end
+
+  def my_none?(pattern = nil)
+    if block_given?
+      my_each { |v| return false if yield(v) }
+    elsif !pattern.nil?
+      if pattern.is_a?(Regexp)
+        my_each { |v| return false if pattern =~ v.to_s }
+      else
+        my_each { |v| return false if v.is_a?(pattern) }
+      end
+    else
+      my_each { |v|  return false if v }
+    end
+    true
+  end
+
+  def my_count(comp = nil)
+    res = 0
+    if block_given?
+      my_each { |v| res += 1 if yield(v) }
+    elsif !comp.nil?
+      my_each { |v| res += 1 if v == comp }
+    else
+      res = size
+    end
+    res
+  end
+
+  def my_map(proc = nil)
+    res = []
+    return to_enum(__method__) unless block_given?
+    if proc.nil? 
+      my_each { |v| res.push(yield v) }
+    else
+      my_each { |v| res.push(proc.call(v)) }
+    end
+    res
+  end
+
+  def my_inject(arg = nil)
```

Comment:
This is actually hardcoding, and it should be avoided whenever possible. Inject method in ruby is working in a way that is always dealing with two numbers: the accumulator (the sum of previous numbers) and the current value, the accumulator is always starting as 0 (for sum/subtraction) or 1(for multiplication/division). Array [1,2,3,4,5] ,for example, will work like this accumulator = 0, next = 1 , 0+1 = 1 then, accumulator is 1 and next value is 2 so 1+2 =3, and so on till we reach 15 in this case. In your case, you are taking two values right away, and making next one index + 1, which is working but not good practice.